pnpm not work with different binary name

### aqua info

```console
$ aqua info
{
  "version": "2.48.3",
  "commit_hash": "ac88862ea6a47c06b44b3fa0017ba541aca6708b",
  "os": "linux",
  "arch": "amd64",
  "pwd": "/home/(USER)/src/github.com/<org>/<repo>",
  "root_dir": "/home/(USER)/.local/share/aquaproj-aqua",
  "env": {
    "AQUA_GLOBAL_CONFIG": "/home/(USER)/.config/aquaproj-aqua/aqua.yaml",
    "AQUA_POLICY_CONFIG": "/home/(USER)/.config/aquaproj-aqua/aqua-policy.yaml"
  },
  "config_files": [
    {
      "path": "/home/(USER)/src/github.com/<org>/<repo>/aqua.yaml"
    }
  ]
}
```

### Overview

Since pnpm v10.6.2, pnpm is aware of the executable name it was invoked with.
When pnpm is installed via [aqua](https://github.com/aquaproj/aqua), the binary name becomes `pnpm-linux-x64,` which causes the ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL error the first time you install a project-specific version of pnpm.

https://github.com/pnpm/pnpm/commit/0b0bcfa143de625135fad76d7a33ea7dfd174b5c

Renaming the binary from pnpm-linux-x64 to pnpm prevents this error.

### How to reproduce

aqua.yaml

```yaml
---
# aqua - Declarative CLI Version Manager
# https://aquaproj.github.io/
# checksum:
#   enabled: true
#   require_checksum: true
#   supported_envs:
#   - all
registries:
- type: standard
  ref: v4.245.1 # renovate: depName=aquaproj/aqua-registry
packages:
- name: pnpm/pnpm@v10.6.2
```

package.json

```json
{
  "name": "frontend",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "engines": {
    "node": "20.10.0"
  },
  "packageManager": "pnpm@9.15.3"
}
```

Executed command and output

```console
$ pnpm install
 ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL  Command "/snapshot/dist/pnpm.cjs" not found
 ERROR  Command failed with exit code 254: /home/<user>/.local/share/aquaproj-aqua/pkgs/github_release/github.com/pnpm/pnpm/v10.6.2/pnpm-linux-x64/pnpm-linux-x64 /snapshot/dist/pnpm.cjs add @pnpm/linux-x64@9.15.3 --loglevel=error --allow-build=@pnpm/exe --config.node-linker=hoisted --config.bin=/home/<user>/.local/share/pnpm/.tools/@pnpm+linux-x64/9.15.3_tmp_2586430/bin
```

### Expected behaviour

The binary should be named `pnpm`, or `process.execPath` should be spoofed to appear as `pnpm`.

### Actual behaviour

process.execPath is likely set to `pnpm-linux-x64`, this cause unexpected behavior in pnpm.

### Note

_No response_

Hmm. `files[].link` doesn't solve this issue.

```yaml
        files:
          - name: pnpm
            link: pnpm
```


```console
$ ls -lh /Users/shunsukesuzuki/.local/share/aquaproj-aqua/pkgs/github_release/github.com/pnpm/pnpm/v10.8.1/pnpm-macos-arm64/
total 62M
lrwxr-xr-x 1 shunsukesuzuki staff  16  4 18 12:29 pnpm -> pnpm-macos-arm64
-rwxr-xr-x 1 shunsukesuzuki staff 62M  4 18 12:29 pnpm-macos-arm64

$ /Users/shunsukesuzuki/.local/share/aquaproj-aqua/pkgs/github_release/github.com/pnpm/pnpm/v10.8.1/pnpm-macos-arm64/pnpm install
 ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL  Command "/snapshot/dist/pnpm.cjs" not found
 ERROR  Command failed with exit code 254: /Users/shunsukesuzuki/.local/share/aquaproj-aqua/pkgs/github_release/github.com/pnpm/pnpm/v10.8.1/pnpm-macos-arm64/pnpm-macos-arm64 /snapshot/dist/pnpm.cjs add @pnpm/macos-arm64@9.15.3 --loglevel=error --allow-build=@pnpm/exe --config.node-linker=hoisted --config.bin=/Users/shunsukesuzuki/Library/pnpm/.tools/@pnpm+macos-arm64/9.15.3_tmp_25157/bin
For help, run: pnpm help install
```

--- a/.github/workflows/wc-integration-test.yaml
+++ b/.github/workflows/wc-integration-test.yaml
@@ -342,7 +342,14 @@ jobs:
       - run: |
           go install -ldflags "-X main.version=v2.48.2-local -X main.commit=$sha -X main.date=$(date +"%Y-%m-%dT%H:%M:%SZ%:z" | tr -d '+')" ./cmd/aqua
 
+      - run: echo "$(aqua root-dir)\bin" >> "$GITHUB_PATH"
+        if: runner.os == 'Windows'
       - run: echo "$(aqua root-dir)/bin" >> "$GITHUB_PATH"
+        if: runner.os != 'Windows'
+      - run: |
+          ("$(aqua root-dir)\bin") | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
+        shell: pwsh
+        if: runner.os == 'Windows'
 
       - run: aqua policy allow
 
@@ -351,7 +358,6 @@ jobs:
         working-directory: tests/pnpm
       - name: Test pnpm (hardlink)
         run: pnpm install
-        shell: bash
         working-directory: tests/pnpm
 
       - name: Test minisign

--- a/.github/workflows/wc-integration-test.yaml
+++ b/.github/workflows/wc-integration-test.yaml
@@ -351,6 +351,7 @@ jobs:
         working-directory: tests/pnpm
       - name: Test pnpm (hardlink)
         run: pnpm install
+        shell: bash
         working-directory: tests/pnpm
 
       - name: Test minisign

--- a/.github/workflows/wc-integration-test.yaml
+++ b/.github/workflows/wc-integration-test.yaml
@@ -354,6 +354,7 @@ jobs:
         working-directory: tests/pnpm
       - name: Test pnpm (hardlink)
         run: pnpm install
+        shell: bash
         working-directory: tests/pnpm
 
       - name: Test minisign

--- a/.github/workflows/wc-integration-test.yaml
+++ b/.github/workflows/wc-integration-test.yaml
@@ -342,10 +342,7 @@ jobs:
       - run: |
           go install -ldflags "-X main.version=v2.48.2-local -X main.commit=$sha -X main.date=$(date +"%Y-%m-%dT%H:%M:%SZ%:z" | tr -d '+')" ./cmd/aqua
 
-      - run: echo "${XDG_DATA_HOME:-$HOME/.local/share}/aquaproj-aqua/bin" >> "$GITHUB_PATH"
-        if: runner.os != 'Windows'
-      - run: echo "${AQUA_ROOT_DIR:-$HOME/AppData/Local/aquaproj-aqua}/bin" >> "$GITHUB_PATH"
-        if: runner.os == 'Windows'
+      - run: echo "$(aqua root-dir)/bin" >> "$GITHUB_PATH"
 
       - run: aqua policy allow
 
@@ -354,7 +351,6 @@ jobs:
         working-directory: tests/pnpm
       - name: Test pnpm (hardlink)
         run: pnpm install
-        shell: bash
         working-directory: tests/pnpm
 
       - name: Test minisign

--- a/.github/workflows/wc-integration-test.yaml
+++ b/.github/workflows/wc-integration-test.yaml
@@ -342,7 +342,7 @@ jobs:
       - run: |
           go install -ldflags "-X main.version=v2.48.2-local -X main.commit=$sha -X main.date=$(date +"%Y-%m-%dT%H:%M:%SZ%:z" | tr -d '+')" ./cmd/aqua
 
-      - run: echo "$(aqua root-dir)\bin" >> "$GITHUB_PATH"
+      - run: printf "%s\\\\bin\n" "$(aqua root-dir)" >> "$GITHUB_PATH"
         if: runner.os == 'Windows'
       - run: echo "$(aqua root-dir)/bin" >> "$GITHUB_PATH"
         if: runner.os != 'Windows'
@@ -359,6 +359,11 @@ jobs:
       - name: Test pnpm (hardlink)
         run: pnpm install
         working-directory: tests/pnpm
+      - name: Test pnpm (hardlink)
+        run: pnpm install
+        working-directory: tests/pnpm
+        shell: bash
+        if: runner.os == 'Windows'
 
       - name: Test minisign
         run: aqua -c aqua.yaml i

--- a/.github/workflows/wc-integration-test.yaml
+++ b/.github/workflows/wc-integration-test.yaml
@@ -32,6 +32,7 @@ jobs:
       - run: aqua --help
       - run: aqua help
       - run: aqua help-all
+
       - run: echo "standard,kubernetes-sigs/kind" | aqua g -f -
       - run: echo "x-motemen/ghq" | aqua g -f -
       - run: echo "local,aquaproj/aqua-installer" | aqua -c tests/main/aqua-global.yaml g -f -
@@ -342,6 +343,13 @@ jobs:
           go install -ldflags "-X main.version=v2.48.2-local -X main.commit=$sha -X main.date=$(date +"%Y-%m-%dT%H:%M:%SZ%:z" | tr -d '+')" ./cmd/aqua
       - run: echo "${XDG_DATA_HOME:-$HOME/.local/share}/aquaproj-aqua/bin" >> "$GITHUB_PATH"
       - run: aqua policy allow
+
+      - name: Test pnpm (hardlink)
+        run: aqua -c aqua.yaml i
+        working-directory: tests/pnpm
+      - name: Test pnpm (hardlink)
+        run: pnpm install
+        working-directory: tests/pnpm
 
       - name: Test minisign
         run: aqua -c aqua.yaml i

--- a/.github/workflows/wc-integration-test.yaml
+++ b/.github/workflows/wc-integration-test.yaml
@@ -344,7 +344,7 @@ jobs:
 
       - run: echo "${XDG_DATA_HOME:-$HOME/.local/share}/aquaproj-aqua/bin" >> "$GITHUB_PATH"
         if: runner.os != 'Windows'
-      - run: echo "${AQUA_ROOT_DIR:-$HOME/AppData/Local/aquaproj-aqua}/bin" >> $GITHUB_PATH
+      - run: echo "${AQUA_ROOT_DIR:-$HOME/AppData/Local/aquaproj-aqua}/bin" >> "$GITHUB_PATH"
         if: runner.os == 'Windows'
 
       - run: aqua policy allow

--- a/.github/workflows/wc-integration-test.yaml
+++ b/.github/workflows/wc-integration-test.yaml
@@ -341,7 +341,12 @@ jobs:
           cache: true
       - run: |
           go install -ldflags "-X main.version=v2.48.2-local -X main.commit=$sha -X main.date=$(date +"%Y-%m-%dT%H:%M:%SZ%:z" | tr -d '+')" ./cmd/aqua
+
       - run: echo "${XDG_DATA_HOME:-$HOME/.local/share}/aquaproj-aqua/bin" >> "$GITHUB_PATH"
+        if: runner.os != 'Windows'
+      - run: echo "${AQUA_ROOT_DIR:-$HOME/AppData/Local/aquaproj-aqua}/bin" >> $GITHUB_PATH
+        if: runner.os == 'Windows'
+
       - run: aqua policy allow
 
       - name: Test pnpm (hardlink)

--- a/aqua/aqua-policy.yaml
+++ b/aqua/aqua-policy.yaml
@@ -3,7 +3,7 @@
 # aqua Policy
 registries:
   - type: standard
-    ref: semver(">= 3.0.0") or Version == "f9cce37273a70e2f5664fb4c3708169ffe7e320c"
+    ref: semver(">= 3.0.0") or Version in ["f9cce37273a70e2f5664fb4c3708169ffe7e320c", "1f20d0c2211df45694e762fcb20830d5c3cedf95"]
   - name: main
     type: local
     path: ../tests/main/registry.yaml

--- a/json-schema/registry.json
+++ b/json-schema/registry.json
@@ -205,6 +205,9 @@
         },
         "link": {
           "type": "string"
+        },
+        "hard": {
+          "type": "boolean"
         }
       },
       "additionalProperties": false,

--- a/pkg/config/registry/config.go
+++ b/pkg/config/registry/config.go
@@ -17,6 +17,7 @@ type File struct {
 	Dir  string `json:"dir,omitempty" yaml:",omitempty"`
 	// Link is the relative path from Src to the link
 	Link string `json:"link,omitempty" yaml:",omitempty"`
+	Hard bool   `json:"hard,omitempty" yaml:",omitempty"`
 }
 
 func (p *PackageInfos) ToMap(logE *logrus.Entry) map[string]*PackageInfo {

--- a/pkg/installpackage/check_file.go
+++ b/pkg/installpackage/check_file.go
@@ -158,6 +158,7 @@ func (is *Installer) createFileLink(logE *logrus.Entry, file *registry.File, exe
 			// do nothing
 			return nil
 		}
+		logE.Info("creating a hard link")
 		if err := is.linker.Hardlink(exePath, link); err != nil {
 			return fmt.Errorf("create a hard link: %w", err)
 		}

--- a/pkg/installpackage/check_file.go
+++ b/pkg/installpackage/check_file.go
@@ -161,6 +161,7 @@ func (is *Installer) createFileLink(logE *logrus.Entry, file *registry.File, exe
 		if err := is.linker.Hardlink(exePath, link); err != nil {
 			return fmt.Errorf("create a hard link: %w", err)
 		}
+		return nil
 	}
 	// file.Link is the relative path from exePath to the link
 	link := filepath.Join(filepath.Dir(exePath), file.Link)

--- a/pkg/installpackage/check_file.go
+++ b/pkg/installpackage/check_file.go
@@ -146,26 +146,30 @@ func (is *Installer) checkFileSrc(ctx context.Context, logE *logrus.Entry, pkg *
 	return exePath, nil
 }
 
+func (is *Installer) createFileHardLink(logE *logrus.Entry, file *registry.File, exePath string) error {
+	link := filepath.Join(filepath.Dir(exePath), file.Link)
+	if is.runtime.IsWindows() && filepath.Ext(link) == "" {
+		link += ".exe"
+	}
+	if f, err := afero.Exists(is.fs, link); err != nil {
+		return fmt.Errorf("check if a hardlink exists: %w", err)
+	} else if f {
+		// do nothing
+		return nil
+	}
+	logE.Info("creating a hard link")
+	if err := is.linker.Hardlink(exePath, link); err != nil {
+		return fmt.Errorf("create a hard link: %w", err)
+	}
+	return nil
+}
+
 func (is *Installer) createFileLink(logE *logrus.Entry, file *registry.File, exePath string) error {
 	if file.Link == "" {
 		return nil
 	}
 	if file.Hard {
-		link := filepath.Join(filepath.Dir(exePath), file.Link)
-		if is.runtime.IsWindows() && filepath.Ext(link) == "" {
-			link += ".exe"
-		}
-		if f, err := afero.Exists(is.fs, link); err != nil {
-			return fmt.Errorf("check if a hardlink exists: %w", err)
-		} else if f {
-			// do nothing
-			return nil
-		}
-		logE.Info("creating a hard link")
-		if err := is.linker.Hardlink(exePath, link); err != nil {
-			return fmt.Errorf("create a hard link: %w", err)
-		}
-		return nil
+		return is.createFileHardLink(logE, file, exePath)
 	}
 	// file.Link is the relative path from exePath to the link
 	link := filepath.Join(filepath.Dir(exePath), file.Link)

--- a/pkg/installpackage/check_file.go
+++ b/pkg/installpackage/check_file.go
@@ -152,6 +152,9 @@ func (is *Installer) createFileLink(logE *logrus.Entry, file *registry.File, exe
 	}
 	if file.Hard {
 		link := filepath.Join(filepath.Dir(exePath), file.Link)
+		if is.runtime.IsWindows() && filepath.Ext(link) == "" {
+			link += ".exe"
+		}
 		if f, err := afero.Exists(is.fs, link); err != nil {
 			return fmt.Errorf("check if a hardlink exists: %w", err)
 		} else if f {

--- a/pkg/installpackage/check_file.go
+++ b/pkg/installpackage/check_file.go
@@ -11,6 +11,7 @@ import (
 	"github.com/aquaproj/aqua/v2/pkg/config/registry"
 	"github.com/aquaproj/aqua/v2/pkg/osfile"
 	"github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
 	"github.com/suzuki-shunsuke/logrus-error/logerr"
 )
 
@@ -148,6 +149,18 @@ func (is *Installer) checkFileSrc(ctx context.Context, logE *logrus.Entry, pkg *
 func (is *Installer) createFileLink(logE *logrus.Entry, file *registry.File, exePath string) error {
 	if file.Link == "" {
 		return nil
+	}
+	if file.Hard {
+		link := filepath.Join(filepath.Dir(exePath), file.Link)
+		if f, err := afero.Exists(is.fs, link); err != nil {
+			return fmt.Errorf("check if a hardlink exists: %w", err)
+		} else if f {
+			// do nothing
+			return nil
+		}
+		if err := is.linker.Hardlink(exePath, link); err != nil {
+			return fmt.Errorf("create a hard link: %w", err)
+		}
 	}
 	// file.Link is the relative path from exePath to the link
 	link := filepath.Join(filepath.Dir(exePath), file.Link)

--- a/pkg/installpackage/check_file.go
+++ b/pkg/installpackage/check_file.go
@@ -168,7 +168,7 @@ func (is *Installer) createFileLink(logE *logrus.Entry, file *registry.File, exe
 	if file.Link == "" {
 		return nil
 	}
-	if file.Hard {
+	if file.Hard || is.runtime.IsWindows() {
 		return is.createFileHardLink(logE, file, exePath)
 	}
 	// file.Link is the relative path from exePath to the link

--- a/pkg/installpackage/link.go
+++ b/pkg/installpackage/link.go
@@ -83,7 +83,7 @@ func (is *Installer) createCmdLink(logE *logrus.Entry, file *registry.File, cmd 
 		})
 	}
 	if is.realRuntime.IsWindows() {
-		if err := is.createHardLink(logE, cmd, aquaProxyPathOnWindows); err != nil {
+		if err := is.createHardLinkToProxy(logE, cmd, aquaProxyPathOnWindows); err != nil {
 			return fmt.Errorf("create a hard link to aqua-proxy: %w", err)
 		}
 		return nil
@@ -94,7 +94,7 @@ func (is *Installer) createCmdLink(logE *logrus.Entry, file *registry.File, cmd 
 	return nil
 }
 
-func (is *Installer) createHardLink(logE *logrus.Entry, cmd string, aquaProxyPathOnWindows string) error {
+func (is *Installer) createHardLinkToProxy(logE *logrus.Entry, cmd, aquaProxyPathOnWindows string) error {
 	hardLink := filepath.Join(is.rootDir, "bin", cmd+".exe")
 	if f, err := afero.Exists(is.fs, hardLink); err != nil {
 		return fmt.Errorf("check if a hard link to aqua-proxy exists: %w", err)

--- a/tests/pnpm/aqua-checksums.json
+++ b/tests/pnpm/aqua-checksums.json
@@ -1,0 +1,34 @@
+{
+  "checksums": [
+    {
+      "id": "github_release/github.com/pnpm/pnpm/v10.8.1/pnpm-linux-arm64",
+      "checksum": "F01B8A3BD14AD93771DBCDD517156A3543E03BFDBCEB63AC159EE3A1D49A978B",
+      "algorithm": "sha256"
+    },
+    {
+      "id": "github_release/github.com/pnpm/pnpm/v10.8.1/pnpm-linux-x64",
+      "checksum": "11B81FBAD6333579AF6B03106D6A4F49CF669E9B96D68DB18E42DB17427F6660",
+      "algorithm": "sha256"
+    },
+    {
+      "id": "github_release/github.com/pnpm/pnpm/v10.8.1/pnpm-macos-arm64",
+      "checksum": "EC94E924D0ED63E273AAE2B9326A5C0E03C6F1883D820241B14EA1F6D9E9900B",
+      "algorithm": "sha256"
+    },
+    {
+      "id": "github_release/github.com/pnpm/pnpm/v10.8.1/pnpm-macos-x64",
+      "checksum": "BF4EABB466CE356547D62D89DF7CEE58DB072D5222CB710897AB8CCA771BB0E0",
+      "algorithm": "sha256"
+    },
+    {
+      "id": "github_release/github.com/pnpm/pnpm/v10.8.1/pnpm-win-x64.exe",
+      "checksum": "E95CB87936D91C6E3D4F21BA0CD2A3E90C6E64F8F477402EBE0FA4DBBF13AA70",
+      "algorithm": "sha256"
+    },
+    {
+      "id": "registries/github_content/github.com/aquaproj/aqua-registry/1f20d0c2211df45694e762fcb20830d5c3cedf95/registry.yaml",
+      "checksum": "FB9BB917AFFB0BFC808094EBBEE1EF8EE55B3095763FDC5EBBE0BD8EBAA00DEC5143BAE9578DE9977271ECD99BD88D961F6DA21617CF5E73638F2572BCA34A64",
+      "algorithm": "sha512"
+    }
+  ]
+}

--- a/tests/pnpm/aqua.yaml
+++ b/tests/pnpm/aqua.yaml
@@ -1,0 +1,12 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/aquaproj/aqua/main/json-schema/aqua-yaml.json
+# aqua - Declarative CLI Version Manager
+# https://aquaproj.github.io/
+checksum:
+  enabled: true
+  require_checksum: true
+registries:
+  - type: standard
+    ref: 1f20d0c2211df45694e762fcb20830d5c3cedf95 # renovate: depName=aquaproj/aqua-registry
+packages:
+  - name: pnpm/pnpm@v10.8.1

--- a/tests/pnpm/package.json
+++ b/tests/pnpm/package.json
@@ -1,0 +1,10 @@
+{
+  "name": "frontend",
+  "private": true,
+  "version": "0.0.0",
+  "type": "module",
+  "engines": {
+    "node": "20.10.0"
+  },
+  "packageManager": "pnpm@9.15.3"
+}

